Update functions to support recording

timmydoza opened the pull request:
<!-- Describe your Pull Request. You may remove some parts that are not applicable. -->

**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Pull Request Details

### Description

This PR adds functionality to the serverless functions to add support for recording. It also refactors the error handling logic to simplify the functions and remove much repetitive code.

To support recording, we create a new Sync Document called the "stream document", which contains information about the state of the stream recording. It can also contain more information down the road if needed.

This stream document is created when a [host creates a stream](https://github.com/twilio/twilio-live-interactive-video/pull/111/commits/71e6a95c33c1762629413ea19f2318da6b2c44a0#diff-afa617e9349b3f351ce76d266070c8e83247f405c623458ac3a1a37956dfabe7R73-R90). The `is_recording` boolean in the stream document is set to the value of the `record_stream` parameter that is a part of the request. 

Note: recording isn't started in the create-stream function. This will come in a later PR. 

This PR also [adds a fake](https://github.com/twilio/twilio-live-interactive-video/pull/111/commits/816d80db2fef6e9de26e0b5dd9a00c70291fc96b) `media-processor-webhook` function. This function will eventually be registered as a media processor webhook, but for now it can be used to trigger a fake recording error in the stream-document. This is useful to test the error reporting UI in the apps, for instance, if you visit this endpoint in your browser:

`https://twilio-live-interactive-video-xxxx-yyyy-dev.twil.io/media-processor-webhook?stream_name=test-stream&recording_error=true`

It will save an error message to the stream-document, which can then trigger an error message in the apps.

This PR also contains a significant refactor of all error handling. It's quite a lot of code to review, but it greatly reduces the size of each file while improving its readability. 

## Burndown

### Before review
* [ ] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with all effected platforms
* [ ] Manually sanity tested running locally
* [ ] Included screenshot as PR comment (if needed)
* [ ] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Re-tested if necessary

timrozum reviewed (approved):
Looks like the new error handling will make the code a lot easier to follow!

olipyskoty commented on serverless/functions/media-processor-webhook.js:
```diff
@@ -0,0 +1,33 @@
+
+exports.handler = async function (context, event, callback) {
+  const common = require(Runtime.getAssets()['/common.js'].path);
+  const { axiosClient, response, getStreamMapItem, createErrorHandler } = common(context, event, callback);
```
are `axiosClient` and `response` being used?

olipyskoty commented on serverless/functions/media-processor-webhook.js:
```diff
@@ -0,0 +1,33 @@
+    const doc = await streamSyncClient.documents('stream').fetch();
+    await streamSyncClient
+      .documents(doc.sid)
+      .update({ data: { ...doc.data, recording: { is_recording: false, error: 'A recording error has occured' } } })
```
should be `occurred`🙂

olipyskoty commented on serverless/functions/join-stream-as-speaker.js:
```diff
@@ -48,156 +47,80 @@ module.exports.handler = async (context, event, callback) => {
+    .catch((error) => {
+      // Ignore "Participant already exists" error (50433)
+      if (error.code !== 50433) {
+        createErrorHandler('error adding read access to user document')(error);
```
Should this error message explain that there was an error adding the participant to the conversation?  i.e. `'error creating conversation participant'`

olipyskoty reviewed (changes requested):
Looks good, and a lot more readable! Found a couple of typos, otherwise looks great.

timmydoza commented:
Good eye @olipyskoty! I fixed those issues. 

@timrozum - I also went ahead and removed the `sync_object_names` from the responses as we discussed.

olipyskoty reviewed (approved):
🎉
